### PR TITLE
Add some hooks to StateMachineManager and NodeSchedulerService

### PR DESCRIPTION
**Pull request by mnesbit**
I have added some hooks to the NodeSchedulerService, StateMachineManager and InMemoryNetwork so that unit tests with scheduled actions can pause using net.waitQuiescent() until all node activity has finally ended. I have also added a test which uses a SchedulableState to trigger a flow and move states to another node using the FinalityFlow, so that we actually exercise this activity more than in the rather specialist scheduler tests.

### CODE REVIEW

File: node/src/test/kotlin/net/corda/node/services/ScheduledFlowTests.kt [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+import java.time.Instant
+import kotlin.test.assertEquals
+
+class ScheduledFlowTests {
+    lateinit var net: MockNetwork
+    lateinit var notaryNode: MockNetwork.MockNode
+    lateinit var nodeA: MockNetwork.MockNode
+    lateinit var nodeB: MockNetwork.MockNode
+
+    data class ScheduledContract(override val legalContractReference: SecureHash = SecureHash.sha256("test")) : Contract {
```
Comment (mikehearn) [inline node/src/test/kotlin/net/corda/node/services/ScheduledFlowTests.kt]: There's actually a DummyContract that works this way already.

File: node/src/main/kotlin/net/corda/node/internal/AbstractNode.kt [hunk 2]
```diff
@@ -176,7 +176,7 @@ abstract class AbstractNode(open val configuration: NodeConfiguration, val netwo
         get() = _networkMapRegistrationFuture
 
     /** Fetch CordaPluginRegistry classes registered in META-INF/services/net.corda.core.node.CordaPluginRegistry files that exist in the classpath */
-    val pluginRegistries: List<CordaPluginRegistry> by lazy {
+    open val pluginRegistries: List<CordaPluginRegistry> by lazy {
```
Comment (rick-r3) [inline node/src/main/kotlin/net/corda/node/internal/AbstractNode.kt]: A good start, but unfortunately if you look at line 116 of RPCStructures it resolves the plugins again.  I suspect we'll have to tunnel the plugins from the node all the way through to there.

Comment (rick-r3) [inline node/src/main/kotlin/net/corda/node/internal/AbstractNode.kt]: Actually, as just discussed, the RPC is not enabled for the mocks, so I think we get away with it.

File: test-utils/src/main/kotlin/net/corda/testing/node/MockNode.kt [hunk 3]
```diff
@@ -281,4 +288,18 @@ class MockNetwork(private val networkSendManuallyPumped: Boolean = false,
         require(nodes.isNotEmpty())
         nodes.forEach { if (it.started) it.stop() }
     }
+
+    // Test method to block until all scheduled activity, active flows
+    // and network activity has ceased.
+    fun waitQuiescent() {
+        while(nodes.any { it.smm.unfinishedFibers.count > 0
```
Comment (exFalso) [inline test-utils/src/main/kotlin/net/corda/testing/node/MockNode.kt]: Isn't there a race condition here? What if we have nodes `[A, B, C, D .. Z]`, we start traversing them with `any`, reach say `C`, now `Z` sends something to `A`

Comment (mnesbit) [inline test-utils/src/main/kotlin/net/corda/testing/node/MockNode.kt]: As flows exist for the lifetime of the action in practice there is always an overlap of objects lifetimes either in the scheduler, or the statemachine, or the network queues. I agree it may still have an edge condition, but I don't have a better solution at this time.

Comment (mnesbit) [inline test-utils/src/main/kotlin/net/corda/testing/node/MockNode.kt]: Yes thinking more I agree there is always a possible ordering in which they can work around my scanning, but I don't have a better solution. Or at least not without forcing locks across all nodes in the simulated network.
